lib/spack/spack/database.py

```diff
@@ -0,0 +1,290 @@
+# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+import os
+import sys
+import inspect
+import glob
+import imp
+
+import time
+import copy
+import errno
+
+from external import yaml
+from external.yaml.error import MarkedYAMLError
+
+import llnl.util.tty as tty
+from llnl.util.filesystem import join_path
+from llnl.util.lang import *
+
+import spack.error
+import spack.spec
+from spack.spec import Spec
+from spack.error import SpackError
+from spack.virtual import ProviderIndex
+from spack.util.naming import mod_to_class, validate_module_name
+
+
+def _autospec(function):
+    """Decorator that automatically converts the argument of a single-arg
+       function to a Spec."""
+    def converter(self, spec_like, **kwargs):
+        if not isinstance(spec_like, spack.spec.Spec):
+            spec_like = spack.spec.Spec(spec_like)
+        return function(self, spec_like, **kwargs)
+    return converter
+
+
+class Database(object):
+    def __init__(self,root,file_name="specDB.yaml"):
+        """
+        Create an empty Database
+        Location defaults to root/specDB.yaml
+        The individual data are dicts containing
+        spec: the top level spec of a package
+        path: the path to the install of that package
+        dep_hash: a hash of the dependence DAG for that package
+        """
+        self.root = root
+        self.file_name = file_name
+        self.file_path = join_path(self.root,self.file_name)
+
+        self.lock_name = "db_lock"
+        self.lock_path = join_path(self.root,self.lock_name)
+
+        self.data = []
+        self.last_write_time = 0
+
+
+    def from_yaml(self,stream):
+        """
+        Fill database from YAML, do not maintain old data
+        Translate the spec portions from node-dict form to spec from
+        """
+        try:
+            file = yaml.load(stream)
+        except MarkedYAMLError, e:
+            raise SpackYAMLError("error parsing YAML database:", str(e))
+
+        if file==None:
+            return
+
+        data = {}
+        for index, sp in file['database'].items():
+            spec = Spec.from_node_dict(sp['spec'])
+            deps = sp['dependency_indices']
+            path = sp['path']
+            dep_hash = sp['hash']
+            db_entry = {'deps':deps, 'spec': spec, 'path': path, 'hash':dep_hash}
+            data[index] = db_entry
+
+        for sph in data.values():
+            for idx in sph['deps']:
+                sph['spec'].dependencies[data[idx]['spec'].name] = data[idx]['spec']
+
+        self.data = data.values()
+
+
+    def read_database(self):
+        """
+        Re-read Database from the data in the set location
+        If the cache is fresh, return immediately.
+        Implemented with mkdir locking for the database file.
+        """
+        if not self.is_dirty():
+            return
+
+        lock=0
+        while lock==0:
```

Comment:
What happens if this loop hangs forever?  This needs a timeout.

We might consider adding something like `lockfile` to externals:

https://pypi.python.org/pypi/lockfile

so that we don't reinvent the wheel.